### PR TITLE
8321069: JvmtiThreadState::state_for_while_locked() returns nullptr for an attached JNI thread with a java.lang.Thread object after JDK-8319935

### PR DESCRIPTION
**Pull request by dcubed-ojdk**
In the fix for the following bug:

[JDK-8319935](https://bugs.openjdk.org/browse/JDK-8319935) Ensure only one JvmtiThreadState is created for one JavaThread associated with attached native thread

JvmtiThreadState::state_for_while_locked() was changed to
return nullptr for attaching JNI threads regardless of whether
that JNI thread/JavaThread had a java.lang.Thread object.

We should only filter out a JavaThread that's attaching via JNI
if it has no java.lang.Thread object.

This fix has been tested with Mach5 Tier[1-7] and there are no related test failures.
Mach5 Tier8 is in process.

I'm going to need @jianglizhou to rerun her testing for:

[JDK-8319935](https://bugs.openjdk.org/browse/JDK-8319935) Ensure only one JvmtiThreadState is created for one JavaThread associated with attached native thread

since the test(s) for that fix are not yet integrated in the jdk/jdk repo.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8321069](https://bugs.openjdk.org/browse/JDK-8321069): JvmtiThreadState::state_for_while_locked() returns nullptr for an attached JNI thread with a java.lang.Thread object after JDK-8319935 (**Bug** - P3)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [5c9eb2ec](https://git.openjdk.org/jdk/pull/16934/files/5c9eb2ec3a2425971f1f35e31655ce6c04a473ba)
 * [Jiangli Zhou](https://openjdk.org/census#jiangli) (@jianglizhou - **Reviewer**)
 * [Serguei Spitsyn](https://openjdk.org/census#sspitsyn) (@sspitsyn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16934/head:pull/16934` \
`$ git checkout pull/16934`

Update a local copy of the PR: \
`$ git checkout pull/16934` \
`$ git pull https://git.openjdk.org/jdk.git pull/16934/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16934`

View PR using the GUI difftool: \
`$ git pr show -t 16934`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16934.diff">https://git.openjdk.org/jdk/pull/16934.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16934#issuecomment-1837212349)

### CODE REVIEW

File: src/hotspot/share/prims/jvmtiThreadState.inline.hpp [hunk 1]
```diff
@@ -82,7 +82,9 @@ inline JvmtiThreadState* JvmtiThreadState::state_for_while_locked(JavaThread *th
   JvmtiThreadState *state = thread == nullptr ? nullptr : thread->jvmti_thread_state();
 
   if (state == nullptr && thread != nullptr &&
-      (thread->is_exiting() || thread->is_attaching_via_jni())) {
+      (thread->is_exiting() ||
+       (thread->threadObj() == nullptr && thread->is_attaching_via_jni())
+      )) {
     // Don't add a JvmtiThreadState to a thread that is exiting or is attaching.
     // When a thread is attaching, it may not have a Java level thread object
     // created yet.
```
Comment (dholmes-ora) [inline src/hotspot/share/prims/jvmtiThreadState.inline.hpp:90]: The comment needs adjusting now - suggestion:
```
// Don't add a JvmtiThreadState to a thread that is exiting, or is attaching
// and does not yet have a Java level thread object allocated.
```

**Review by dholmes-ora — APPROVED** [review]
Initially I thought this was not the right fix as we should not be exposing an attaching thread that may still have a partially constructed `threadObj`. But this issue shows that we must expose such a thread because the constructor of the `Thread` object can trigger these events on the current thread so it must have a valid JVMTI state!

Thanks.

File: src/hotspot/share/prims/jvmtiThreadState.inline.hpp (hunk 1 above)
Comment (dcubed-ojdk) [reply src/hotspot/share/prims/jvmtiThreadState.inline.hpp]: Thanks for the suggested change. I agree and I've applied it.

**Comment by dcubed-ojdk** [comment]
@dholmes-ora - Thanks for the review.

**Review by sspitsyn — APPROVED** [review]
Looks good. Thank you for fixing this!

**Comment by dcubed-ojdk** [comment]
> @dcubed-ojdk and @dholmes-ora, is there a case
> JvmtiVMObjectAllocEventCollector::JvmtiVMObjectAllocEventCollector()
> might also see an attaching thread without the thread obj allocated?

JvmtiVMObjectAllocEventCollector is the code path that we ran into with
the internal stress test, but in the case that tripped, we had a thread obj
allocated. If we ran the same code path without the thread obj allocated,
then we would return `nullptr` which was the goal of your original fix.

I don't know if that specific test case is actually reached by any of our tests,
but if such a case occurred, it did not result in the guarantee() firing or any
other failure mode that I saw. The
`guarantee(state != nullptr) failed: exiting thread called setup_jvmti_thread_state`
has not been seen in Mach5 Tier[1-8] testing with this fix in place and I didn't
see any other test failures that could be tracked back to problems with this
code.

A JvmtiVMObjectAllocEventCollector object is created in 34 places in jvm.cpp
alone and I haven't checked all of those. I only checked out the one use in
JVM_GetStackAccessControlContext.

**Comment by dcubed-ojdk** [comment]
@jianglizhou and @sspitsyn - Thanks for the reviews.

In the interest of reducing the noise in the Mach5 CI, I'm going ahead with
integrating this fix without waiting for a reply to my comment above. If there
are remaining issues, then we'll deal with them in a follow-up bug/RFE.

/integrate

**Comment by jianglizhou** [comment]
> In the interest of reducing the noise in the Mach5 CI, I'm going ahead with integrating this fix without waiting for a reply to my comment above. If there are remaining issues, then we'll deal with them in a follow-up bug/RFE.

Thumbs up, thanks!

**Comment by jianglizhou** [comment]
> > @dcubed-ojdk and @dholmes-ora, is there a case
> > JvmtiVMObjectAllocEventCollector::JvmtiVMObjectAllocEventCollector()
> > might also see an attaching thread without the thread obj allocated?
> 
> JvmtiVMObjectAllocEventCollector is the code path that we ran into with the internal stress test, but in the case that tripped, we had a thread obj allocated. If we ran the same code path without the thread obj allocated, then we would return `nullptr` which was the goal of your original fix.
> 
> I don't know if that specific test case is actually reached by any of our tests, but if such a case occurred, it did not result in the guarantee() firing or any other failure mode that I saw. The `guarantee(state != nullptr) failed: exiting thread called setup_jvmti_thread_state` has not been seen in Mach5 Tier[1-8] testing with this fix in place and I didn't see any other test failures that could be tracked back to problems with this code.
> 
> A JvmtiVMObjectAllocEventCollector object is created in 34 places in jvm.cpp alone and I haven't checked all of those. I only checked out the one use in JVM_GetStackAccessControlContext.

Thanks. I was wondering if we would need further work to handle `JvmtiVMObjectAllocEventCollector::JvmtiVMObjectAllocEventCollector()` with something similar to the JvmtiSampledObjectAllocEventCollector change, if it's possible to trigger the `guarantee(state != nullptr)` for attaching thread with no allocated thread obj.

**Comment by dholmes-ora** [comment]
We would need to examine all uses of JvmtiVMObjectAllocEventCollector::JvmtiVMObjectAllocEventCollector() to see if they can occur in the context of an attaching thread prior to allocation of the thread object. I suspect not because we allocate immediately after creating the JavaThread. The issue we hit here was events triggered by the execution of the Java code for the Thread constructor. It may be that we are just lucky that the same Java code doesn't lead to the VMObjectAlloc event though. ???

**Pull request closed without merging**